toolchain/check/handle_impl.cpp

```diff
@@ -210,23 +210,41 @@ static auto PopImplIntroducerAndParamsAsNameComponent(
                               }));
   }
 
-  Parse::NodeId first_param_node_id =
-      context.node_stack().PopForSoloNodeId<Parse::NodeKind::ImplIntroducer>();
+  Parse::Tree::PostorderIterator first_param_iter(
+      context.node_stack().PopForSoloNodeId<Parse::NodeKind::ImplIntroducer>());
   // Subtracting 1 since we don't want to include the final `{` or `;` of the
   // declaration when performing syntactic match.
-  // TODO: Following proposal #3763, we should exclude any `where` clause, and
-  // add `Self` before `as` if needed, see:
-  // https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p3763.md#redeclarations
   auto node_kind = context.parse_tree().node_kind(end_of_decl_node_id);
   CARBON_CHECK(node_kind == Parse::NodeKind::ImplDefinitionStart ||
                node_kind == Parse::NodeKind::ImplDecl);
-  Parse::NodeId last_param_node_id(end_of_decl_node_id.index - 1);
+  Parse::Tree::PostorderIterator last_param_iter(end_of_decl_node_id);
+  --last_param_iter;
+  // Following proposal #3763, exclude a final `where` clause, if present. See:
+  // https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p3763.md#redeclarations
+  node_kind = context.parse_tree().node_kind(*last_param_iter);
+  if (node_kind == Parse::NodeKind::WhereExpr) {
+    int where_operands_to_skip = 1;
+    --last_param_iter;
+    CARBON_CHECK(first_param_iter < last_param_iter);
+    do {
+      node_kind = context.parse_tree().node_kind(*last_param_iter);
+      if (node_kind == Parse::NodeKind::WhereExpr) {
+        // If we have a nested `where`, we need to see another `WhereOperand`
+        // before we find the one that matches our original `WhereExpr` node.
+        ++where_operands_to_skip;
+      } else if (node_kind == Parse::NodeKind::WhereOperand) {
+        --where_operands_to_skip;
+      }
+      --last_param_iter;
+      CARBON_CHECK(first_param_iter < last_param_iter);
+    } while (where_operands_to_skip > 0);
```

Comment:
What do you think about using a `while` to avoid the copied `CHECK`? You could also move in `--last_param_iter;` like so:

```suggestion
    while (where_operands_to_skip > 0) {
      --last_param_iter;
      CARBON_CHECK(first_param_iter < last_param_iter);
      node_kind = context.parse_tree().node_kind(*last_param_iter);
      if (node_kind == Parse::NodeKind::WhereExpr) {
        // If we have a nested `where`, we need to see another `WhereOperand`
        // before we find the one that matches our original `WhereExpr` node.
        ++where_operands_to_skip;
      } else if (node_kind == Parse::NodeKind::WhereOperand) {
        --where_operands_to_skip;
      }
    }
```

Reply:
The problem is that code isn't quite the same since it doesn't skip the final `WhereOperand` node. Which we want so that `Foo where...` compares equal to `Foo` without a `where`. The code that doesn't repeat the decrement and CHECK does:
```
while (true) {
  --last_param_iter;
  CARBON_CHECK(first_param_iter < last_param_iter);
  if (where_operands_to_skip == 0) {
    break;
  }              
  node_kind = context.parse_tree().node_kind(*last_param_iter);
  // ...
```
which isn't shorter, but maybe the lack of repetition makes the code more robust? I'm not sure. Unfortunately C++ doesn't have a convenient control flow construct for loop-and-a-half

Reply:
FWIW I'd lean towards that, although maybe I'm being biased about do/while. Ultimately I don't have a strong preference (although will Carbon have do/while?)